### PR TITLE
[6.2] No longer cache error pages

### PR DESCRIPTION
**Pull request by Chrissi2812**
This prevents error pages from getting stuck in cache. 
If for example one users request caused an error it will no longer affect every user after him.

### CODE REVIEW

**Comment by Chrissi2812** [comment]
@joomla-cms-bot I added the description for this pull request

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.2-dev.

**Comment by Hackwar** [comment]
@Chrissi2812 Please provide instructions how to provoke an error which would be cached and how to then test this PR. I would really like to see this in 5.2, but I can't merge it without tests and people wont test it without instructions. Thank you!

**Comment by weeblr** [comment]
Here are some testing instructions:

1 - Create an are article and link it from a menu item
2 - Verify it's showing normally by visiting the link from the menu item
3 - **Unpublish** the article
3 - Enable the Joomla Cache system plugin
4 - Clear the Joomla cache
5 - Visit the menu item link again: you get a 404 Page not Found error
6 - Publish back the article 

## Without the change

Any visitor keeps getting a 404 when trying to view that page, despite the article having been published again

## With the change

As soon as the article is published again, the article is displayed normally and there's no 404 anymore.

**Comment by brianteeman** [comment]
When you are using the page cache plugin if an article has been cached then it will still be displayed after it has been unpublished unless the cache is cleaned. In other words this PR just exposes a general issue with the page cache plugin (you need to clear the cache after any change) and just solves part of the problem - error pages when cached will still be displayed even when the state of the content has been changed to prevent the error. 
In other words this is a valid issue but the PR only partially resolves it. Surely the correct thing to do is to flush the cache when the content is changed

**Comment by Chrissi2812** [comment]
If I remember it right, this was to address something like an upstream timeout event.

We fetch results from external site and if it's not reachable within 10 seconds it timeouts with an Joomla Error. And then that would be cached for the page Cache duration. 

With the fix instead only the working request would get cached.

**Comment by weeblr** [comment]
Correct. Instructions I gave were a simple way of testing. External sources would be harder to test, for instance a bug in an extension or in the template, causing a 500 or something.

**Comment by brianteeman** [comment]
@Chrissi2812 that usecase makes sense except I wouldnt want to ever cache that page even when it is valid and would set it up as an exception in the plugin  options

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.3-dev.

**Comment by fgsw** [comment]
I have tested this item :white_check_mark: successfully on afc9df609c90439273cf8ab11357ccfd8c0eb439<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42673">issues.joomla.org/tracker/joomla-cms/42673</a>.</sub>

**Comment by exlemor** [comment]
I have tested this item :white_check_mark: successfully on afc9df609c90439273cf8ab11357ccfd8c0eb439<br /><br />Successfully tested during PBF 2025 - February.


However, wouldn't it make sense to make a PR that fixes the situation more globally? 

i.e. IF the Page Cache plugin is turned on, AND a change to content is made that there is an automatic cache cleaning process triggered?

(this would if I understood it correctly correct the situation for all pages not just 404 error page) 

(I apologize in advance if I misunderstood something)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42673">issues.joomla.org/tracker/joomla-cms/42673</a>.</sub>

**Comment by crimle** [comment]
I have tested this item :white_check_mark: successfully on afc9df609c90439273cf8ab11357ccfd8c0eb439<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42673">issues.joomla.org/tracker/joomla-cms/42673</a>.</sub>

**Comment by Chrissi2812** [comment]
@exlemor I did this PR because we had an external Service that would sometimes Rate limit or Timeout. With this change as soon as it comes back online it's all good again.

Clearing all Caches maybe a bit much, but that's just my opinion.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 6.0-dev.

**Comment by richard67** [comment]
I've allowed myself to fix the merge conflict in file "plugins/system/cache/src/Extension/Cache.php" which was caused by PR #44842 after the rebase to 6.0-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 6.1-dev.
